WORK IN PROGRESS: Remove "abi" product flavor and introduce "engine" product flavor.

pocmo opened the pull request:
Incomplete draft. DO NOT MERGE yet. Opening PR to test on taskcluster and share with @mitchhentges.

~~Depends on #4297 (first commit in this PR)~~ Landed ✅

----

This patch will allow us to build Fenix against GeckoView Nightly and GeckoView Beta by
introducing a new flavor dimension: `engine = [geckoNightly, geckoBeta]`.

In addition to that it adds a `fenix`  prefix to the nightly, beta and production flavors
to reduce the ambiguity between Fenix beta/nightly and GeckoView beta/nightly.

For now the build types have the following engine variants enabled:

**debug**: `geckoNightly, geckoBeta`
Both variants enabled for local development and testing.

**forPerformanceTest**: `geckoNightly, geckoBeta`
Both variants enabled unless the perf team only cares about Nightly (tbd)

**fenixNightlyLegacy**: `geckoBeta`
Uses GeckoView Beta for now - the same version we ship production builds with (same behavior
as before). This release type will eventualyl be decommissioned once we switch to a separate
Nightly app on Google Play.

**fenixNightly**: `geckoBeta`
Uses GeckoView Beta for now - the same version we ship production builds with (same behavior
as before). Changing this build to use GeckoView Nightly is currently being discussed.

**fenixBeta**: `geckoBeta`
Fenix Beta uses GeckoView Beta.

**fenixProduction**: `geckoBeta`
Fenix Production uses GeckoView Beta (69) currently.

----

AC provides a mechanism to write feature flags that depend on a specific engine version. I'll write a wiki page or email explaining that once this gets closer to landing.

pocmo commented:
@mitchhentges 

Oh, no, that thing is back:

> expected one element but was: <BuildOutput{apkData=DefaultApkData(_type=FULL_SPLIT, _filters=[FilterData{type=ABI, value=arm64-v8a}], _versionCode=1, _versionName=1.0.1930, _filterName=arm64-v8a, _outputFileName=app-geckoBeta-arm64-v8a-debug.apk, _fullName=geckoBetaArm64-v8aDebug, _baseName=geckoBeta-arm64-v8a-debug, _enabled=true), path=/opt/fenix/app/build/intermediates/processed_res/geckoBetaDebug/processGeckoBetaDebugResources/out/resources-geckoBetaArm64-v8aDebug.ap_, properties=packageId=org.mozilla.fenix.debug,split=,minSdkVersion=21}, BuildOutput{apkData=DefaultApkData(_type=FULL_SPLIT, _filters=[FilterData{type=ABI, value=armeabi-v7a}], _versionCode=1, _versionName=1.0.1930, _filterName=armeabi-v7a, _outputFileName=app-geckoBeta-armeabi-v7a-debug.apk, _fullName=geckoBetaArmeabi-v7aDebug, _baseName=geckoBeta-armeabi-v7a-debug, _enabled=true), path=/opt/fenix/app/build/intermediates/processed_res/geckoBetaDebug/processGeckoBetaDebugResources/out/resources-geckoBetaArmeabi-v7aDebug.ap_, properties=packageId=org.mozilla.fenix.debug,split=,minSdkVersion=21}, BuildOutput{apkData=DefaultApkData(_type=FULL_SPLIT, _filters=[FilterData{type=ABI, value=x86}], _versionCode=1, _versionName=1.0.1930, _filterName=x86, _outputFileName=app-geckoBeta-x86-debug.apk, _fullName=geckoBetaX86Debug, _baseName=geckoBeta-x86-debug, _enabled=true), path=/opt/fenix/app/build/intermediates/processed_res/geckoBetaDebug/processGeckoBetaDebugResources/out/resources-geckoBetaX86Debug.ap_, properties=packageId=org.mozilla.fenix.debug,split=,minSdkVersion=21}, BuildOutput{apkData=DefaultApkData(_type=FULL_SPLIT, _filters=[FilterData{type=ABI, value=x86_64}], _versionCode=1, _versionName=1.0.1930, _filterName=x86_64, _outputFileName=app-geckoBeta-x86_64-debug.apk, _fullName=geckoBetaX86_64Debug, _baseName=geckoBeta-x86_64-debug, _enabled=true), path=/opt/fenix/app/build/intermediates/processed_res/geckoBetaDebug/processGeckoBetaDebugResources/out/resources-geckoBetaX86_64Debug.ap_, properties=packageId=org.mozilla.fenix.debug,split=,minSdkVersion=21}>

pocmo commented:
[Looks like](https://issuetracker.google.com/issues/130740393) this goes away with Android Gradle plugin 3.5 (But there's only a first RC build available); Passes locally with that.

mitchhentges commented:
Interesting! A couple notes:

* That build error can be worked around in the short term by doctoring the `gradle.properties` file, as [done in `reference-browser](https://github.com/mozilla-mobile/reference-browser/issues/840)
* To reduce moving parts, I'd like to successfully land #4297 before we merge this PR
* Aren't we going to be permanently tied to `geckoview` as our engine? Perhaps we can simplify the variant naming to `gecko = ['nightly', 'beta']` :thinking:

pocmo commented:
>     * That build error can be worked around in the short term by doctoring the `gradle.properties` file, as [done in `reference-browser](https://github.com/mozilla-mobile/reference-browser/issues/840)

I wonder how well `android.enableUnitTestBinaryResources=false` is going to work here since Fenix actually has unit tests. However if it works this may be a good temporary solution until version 3.5 of the plugin is released.

>     * To reduce moving parts, I'd like to successfully land #4297 before we merge this PR

Yes, absolutely, I want that too.

>     * Aren't we going to be permanently tied to `geckoview` as our engine? Perhaps we can simplify the variant naming to `gecko = ['nightly', 'beta']` 🤔

Yes, we are. But since the "dimension name" is not part of the build variant and APK names it gets very very confusing if you do not know if it's gecko beta/nightly or fenix beta/nightly.

pocmo commented:
* #4297 landed.
* I rebased this PR.
* I verified that we can set `android.enableUnitTestBinaryResources=false` for now and are tests still pass (we can revert that with Android Gradle plugin 3.5) and updated the PR.
* All PR tasks are now green.

@mitchhentges Is that a state you can take over from to get the publishing part done?

mitchhentges commented:
Yeah, I'll take a look

mitchhentges commented:
Heads up that I might not be able to look at this until next week, I'm focusing on some Amazon Store work atm :smile:

pocmo commented:
>  What's the priority on this? I can shift if this is more critical than I'm interpreting :)

It's not super urgent, but more and more people are asking for Fenix builds using GeckoView Nightly - especially perf and Gecko teams would benefit from that. I want to unblock them and this is as far as I can get on my own. :)

mitchhentges commented on app/build.gradle:
```diff
@@ -70,39 +71,62 @@ android {
+        // | debug              | ✅            | ✅       | Both variants for testing and development.
+        // | forPerformanceTest | ✅            | ✅       | Both variants unless the perf team only cares about Nightly (TBD).
+        // | fenixNightlyLegacy | ❌            | ✅       | For now GV Beta - the same version we ship with (TBD). Release type will be decommissioned soon.
+        // | fenixNightly       | ❌            | ✅       | For now GV Beta - the same version we ship with (TBD).
```
Or rather, we'll be building `fenixNightly` with both `gv-nightly` (in the deployed version of the Fenix Nightly) and `gv-beta` (only uploaded to Taskcluster to be used by QA).
I'll update the contraints as part of my modifications to this PR

pocmo replied:
Yes, we'll be building Fenix Nightly with GeckoView Nightly. At the time I started this PR there was no decision yet.

sblatz commented:
Any update on this @mitchhentges?

mitchhentges commented:
Working on it more today, currently reworking how artifacts flow through the release workflow

pocmo commented:
Closing in favor of #4851
